### PR TITLE
Spatial corr

### PR DESCRIPTION
**Pull request by lcnature**
Update BRSA model to include spatial correlation in noise
some parameter naming changes and bug fixing.

### CODE REVIEW

File: tests/reprsimil/test_brsa.py [hunk 1]
```diff
@@ -110,51 +117,75 @@ def test_fit():
...
-    brsa = BRSA(GP_space=True,GP_inten=True,verbose=False,n_iter = 200)
+    brsa = BRSA(GP_space=True,GP_inten=True,verbose=False,n_iter = 200,auto_nuiance=False)
 
-    brsa.fit(X=Y, design=design.design_used, scan_onsets=scan_onsets,
+    # We also test that it can detect baseline regressor included in the design matrix for task conditions
+    wrong_design = np.insert(design.design_task, 0, 1, axis=1)
+    with pytest.raises(ValueError) as excinfo:
+        brsa.fit(X=Y, design=wrong_design, scan_onsets=scan_onsets,
+             coords=coords, inten=inten)
+    assert 'Your design matrix appears to have included baseline time series.' in str(excinfo.value)
```
Comment (mihaic) [inline tests/reprsimil/test_brsa.py]: The `pytest.raises` above will fail the test if a `ValueError` exception is **not** raised. Therefore, there is no need for this assertion, unless you are concerned a different `ValueError` is raised, unrelated to the design matrix including baseline time series.

File: tests/reprsimil/test_brsa.py [hunk 2]
```diff
@@ -110,51 +117,75 @@ def test_fit():
...
+    p = scipy.stats.spearmanr(u_b[np.tril_indices_from(u_b)],u_i[np.tril_indices_from(u_i)])[1]
     assert p < 0.01, "Fitted covariance matrix does not correlate with ideal covariance matrix!"
     # check that the recovered SNRs makes sense
     p = scipy.stats.pearsonr(brsa.nSNR_,snr)[1]
     assert p < 0.01, "Fitted SNR does not correlate with simualted SNR!"
     assert np.isclose(np.mean(np.log(brsa.nSNR_)),0), "nSNR_ not normalized!"
+    p = scipy.stats.pearsonr(brsa.sigma_,noise_level)[1]
+    assert p < 0.01, "Fitted noise level does not correlate with simualted noise level!"
+    p = scipy.stats.pearsonr(brsa.rho_,rho1)[1]
+    assert p < 0.01, "Fitted AR(1) coefficient does not correlate with simualted values!"
```
Comment (mihaic) [inline tests/reprsimil/test_brsa.py]: Typo in "simualted".

File: tests/reprsimil/test_brsa.py [hunk 3]
```diff
@@ -231,113 +264,219 @@ def test_gradient():
...
+        'Dimension of X0TX0 etc. returned from _prepare_data is wrong'
+    assert np.ndim(XTX0) == 2 and np.ndim(XTDX0) == 2 and np.ndim(XTFX0) == 2,\
+        'Dimension of XTX0 etc. returned from _prepare_data is wrong'
+    assert np.ndim(X0TY) == 2 and np.ndim(X0TDY) == 2 and np.ndim(X0TFY) == 2,\
+        'Dimension of X0TY etc. returned from _prepare_data is wrong'
+    X0 = np.ones(n_T)
     l_idx = np.tril_indices(n_C)
+    # rank = n_C - 1
+    # idx_rank = np.where(l_idx[1] < rank)
+    # l_idx = (l_idx[0][idx_rank], l_idx[1][idx_rank])
```
Comment (mihaic) [inline tests/reprsimil/test_brsa.py]: Please do not comment out unnecessary code; instead, remove it. There are other instances.

**Comment by mihaic** [comment]
@lcnature I think the test failures are caused by the merge conflict in `tests/utils/test_utils.py`. Please merge the latest changes from master and resolve the conflict.

File: tests/reprsimil/test_brsa.py (hunk 3 above)
Comment (lcnature) [inline tests/reprsimil/test_brsa.py]: Fixed.

**Comment by lcnature** [comment]
@mihaic The failure was because of a change after the last time I tested the code. The bug was fixed. But what is the procedure for dealing with the conflict?

**Comment by mihaic** [comment]
@lcnature To fix the conflict, merge the master into your branch, then edit `tests/utils/test_utils.py` to your liking and remove the conflict markers. See the Git documentation for more details:
https://git-scm.com/book/en/v2/Git-Branching-Basic-Branching-and-Merging#Basic-Merge-Conflicts

**Comment by lcnature** [comment]
@mihaic I think the conflict is solved now.

**Comment by mihaic** [comment]
@mshvartsman Given that you reviewed the BRSA code, could you please look at this PR?

**Comment by mshvartsman** [comment]
As per an email exchange, as long as tests pass (finite diffs match hand-written derivatives, and
synthetic estimation looks good), I'm happy to approve this as-is and let subtler bugs shake out as this get used.

**Pull request merged**